<a href="https://colab.research.google.com/github/xDanilog27/Bussiness-Inteligent/blob/main/Reto_1_Modelo_de_Clasificaci%C3%B3_Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**0. Cargar las librerias**

In [ ]:
import numpy as np #Posee Funciones Matematicas
import pandas as pd #Comunicacion excel
from sklearn.metrics import accuracy_score # comparar las predicciones del modelo

**1. Cargar la base de datos**

In [ ]:
nxl='/content/1. BD2_CarCrash_Int.xlsx'
XDB= pd.read_excel(nxl,sheet_name=0)
XDB.head(10)

# Seleccionar datos para el condado de San Francisco (asumiendo que existe una columna 'County')

XDB_SF=XDB[XDB['County'] == 'SAN FRANCISCO'] # Filtrar los datos para San Francisco

# Mostrar los primeros registros de los datos de San Francisco
print(XDB_SF.head(10))

# 1. Convertir 'CrashType' a 'category' y asignar códigos numéricos
XDB_SF['CrashType'] = pd.Categorical(XDB_SF['CrashType'], categories=sorted(XDB_SF['CrashType'].unique()), ordered=True).codes

# 2. Mostrar los datos con la columna 'CrashType' transformada
print(XDB_SF.head(10))

# Separar variables de entrada (XD) y variable objetivo (yd)
XD=XDB_SF[['Weekday', 'ViolCat', 'Daylight', 'Month', 'CrashType', 'ClearWeather']]
yd=XDB_SF['Highway']


        ID         County           City  Weekday  Severity  ViolCat  \
907    908  SAN FRANCISCO  SAN FRANCISCO        7         1        1   
1580  1581  SAN FRANCISCO  SAN FRANCISCO        2         0        9   
1581  1582  SAN FRANCISCO  SAN FRANCISCO        7         0        3   
1586  1587  SAN FRANCISCO  SAN FRANCISCO        7         0        3   
1587  1588  SAN FRANCISCO  SAN FRANCISCO        3         0        3   
1588  1589  SAN FRANCISCO  SAN FRANCISCO        7         0       12   
1607  1608  SAN FRANCISCO  SAN FRANCISCO        7         0       10   
1616  1617  SAN FRANCISCO  SAN FRANCISCO        2         0        9   
1617  1618  SAN FRANCISCO  SAN FRANCISCO        6         0       11   
1621  1622  SAN FRANCISCO  SAN FRANCISCO        7         0        9   

      ClearWeather  Month CrashType  Highway  Daylight  
907              1      3         A        1         1  
1580             1      2         A        0         0  
1581             1      1         A 

<ipython-input-4-5f9c94caf098>:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  XDB_SF['CrashType'] = pd.Categorical(XDB_SF['CrashType'], categories=sorted(XDB_SF['CrashType'].unique()), ordered=True).codes


**2. Implementacion del modelo**

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB() #Aloja el Modelo
mnb.fit(XD,yd) #Entrena el Modelo

#Obtener informacion del modelo
u=mnb.theta_ #Media de los datos
print("La media de los datos es:\n",u) #La primera es No Y la Segunda es Si, en base al Highway

varianza=mnb.var_ #Varianza de los datos
print("La varianza de los datos es:\n",varianza)

sigma=np.sqrt(varianza) #Desviacion Estandar - Se obtiene la desviacion estandar de los ingresos
print("La desviacion estandar de los datos es:\n",sigma)

# Limites superior y limite inferior - NO SE SI DEJAR EL 3 o NO PONER NADA.
LS=u+sigma
LI=u-sigma
print("El limite superior es:\n",LS)
print("El limite inferior es:\n",LI)

XDS=[[5,1,1,3,4,0]]
ydp=mnb.predict(XDS)
print("La prediccion es:\n",ydp)

if ydp==0:
  print("No hay accidente automovilístico en la autopista")
else:
  print("Hay accidente automovilístico en la autopista")



La media de los datos es:
 [[3.88417722 7.45696203 0.66265823 4.35189873 3.21772152 0.86265823]
 [3.9261477  4.1497006  0.62075848 6.69461078 2.27744511 0.79840319]]
La varianza de los datos es:
 [[ 3.80493953 12.40004648  0.22354231  5.49135717  3.5108252   0.11847902]
 [ 4.06440613  4.67419653  0.2354174  11.84486119  1.11863301  0.16095555]]
La desviacion estandar de los datos es:
 [[1.95062542 3.52136997 0.47280262 2.3433645  1.87371962 0.34420782]
 [2.01603723 2.16198902 0.48519831 3.44163641 1.05765448 0.40119266]]
El limite superior es:
 [[ 5.83480263 10.978332    1.13546084  6.69526323  5.09144114  1.20686605]
 [ 5.94218494  6.31168962  1.10595679 10.13624719  3.33509959  1.19959585]]
El limite inferior es:
 [[1.9335518  3.93559205 0.18985561 2.00853424 1.3440019  0.51845041]
 [1.91011047 1.98771158 0.13556017 3.25297437 1.21979063 0.39721054]]
La prediccion es:
 [0]
No hay accidente automovilístico en la autopista


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


In [ ]:
#1. Numero de datos de la base de datos original
total_data = len(XDB)
print(f"Número total de datos de toda la base de datos: {total_data}")

#1. Numero de datos de la base de datos filtrada
total_data = len(XDB_SF)
print(f"Número total de datos: {total_data}")


#2.	¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?

highway_counts=XDB_SF['Highway'].value_counts()
percentage_highway=(highway_counts[1] / total_data) * 100
percentage_no_highway=(highway_counts[0] / total_data) * 100
print(f"Porcentaje de accidentes en autopista: {percentage_highway:.2f}%")
print(f"Porcentaje de accidentes fuera de autopista: {percentage_no_highway:.2f}%")


Número total de datos de toda la base de datos: 112660
Número total de datos: 2081
Porcentaje de accidentes en autopista: 24.07%
Porcentaje de accidentes fuera de autopista: 75.93%


In [ ]:
from sklearn.metrics import confusion_matrix
#Pronostico
ybp2=mnb.predict(XD)
cm=confusion_matrix(yd,ybp2) #Crea matriz de confusion
print("La matriz de confusion es:\n", cm)
VN=cm[0,0] #Verdaderos negativos
VP=cm[1,1] #Verdaderos positivos
FN=cm[1,0] #Falsos negativos
FP=cm[0,1] #Falsos positivos

#6
Exactitud=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud es:\n",Exactitud)

#7
Especificidad=VN/(VN+FP)
print("La especificidad es:\n",Especificidad)

#8 y 9
print(f"Verdaderos Negativos: {VN}")
print(f"Verdaderos Positivos: {VP}")

#10
TasaError=(FP+FN)/(VP+VN+FP+FN)
print("La tasa de error es:\n",TasaError)

#11
Sensibilidad=VP/(VP+FN)
print("La sensibilidad es:\n",Sensibilidad)

#12
Precision=VP/(VP+FP)
print("La precición es:\n",Precision)

#13 5,1,1,3,4,0
print("La prediccion es:\n",ydp)

if ydp==0:
  print("No hay accidente automovilístico en la autopista")
else:
  print("Hay accidente automovilístico en la autopista")



La matriz de confusion es:
 [[1365  215]
 [ 224  277]]
La exactitud es:
 0.7890437289764536
La especificidad es:
 0.8639240506329114
Verdaderos Negativos: 1365
Verdaderos Positivos: 277
La tasa de error es:
 0.21095627102354636
La sensibilidad es:
 0.5528942115768463
La precición es:
 0.5630081300813008
La prediccion es:
 [0]
No hay accidente automovilístico en la autopista


**Análisis de los resultados**

El análisis se centra en la predicción de accidentes automovilísticos en autopistas, utilizando un modelo Naive Bayes Gaussiano y datos solo del condado de San Francisco. La base de datos original fue filtrada, resultando en un conjunto de 2081 accidentes. Un paso crucial durante el proceso fue la conversión de la variable categórica "CrashType" a una representación numérica ordinal, necesaria para el modelo.

La variable objetivo, "Highway" (accidente en autopista o no), presenta un desbalance significativo: solo el 24.07% de los accidentes ocurrieron en autopistas, contrastando con el 75.93% que no

En cuanto al rendimiento, el modelo alcanza una exactitud del 78.9%, clasificando correctamente casi el 8 de cada 10 accidentes. Sin embargo, esta cifra esconde un rendimiento desigual en las distintas categorías. La especificidad, que mide la capacidad para identificar correctamente los accidentes fuera de la autopista, es alta, alcanzando el 86.4%. Esto sugiere que el modelo es confiable para predecir cuándo un accidente no ocurrió en una autopista.

la sensibilidad es considerablemente menor, del 55.3%. Esto indica que el modelo tiene dificultades para predecir correctamente los accidentes que sí ocurren en autopistas. La precisión, que mide la proporción de accidentes clasificados como ocurridos en autopistas que realmente lo fueron, se sitúa en el 56.3%.

El análisis de la matriz de confusión revela que el modelo clasifica correctamente 1365 accidentes como no ocurridos en autopistas (verdaderos negativos) y 277 como ocurridos en autopistas (verdaderos positivos). Sin embargo, también comete errores, clasificando 224 accidentes ocurridos en autopistas como no ocurridos (falsos negativos) y 215 accidentes no ocurridos en autopistas como ocurridos (falsos positivos).

Finalmente, el modelo predice que un accidente con las características [5, 1, 1, 3, 4, 0], no ocurrió en una autopista.

